File: 暑期比赛专用/contest.py
tmp = input()
cin = tmp.split()
n = int(cin[0])
k = int(cin[1])
tmp = input()
cin = tmp.split()
a = []
res = 0
two = 1
for i in range(n+1):
    num = int(cin[i])
    a.append(num)
    res += num * two
    two *= 2
ans = 0
two = 1
for i in range(n+1):
    tmp = res - a[i] * two
    if tmp % two != 0:
        continue
    tmp /= two
    if tmp > 0 and tmp <= k:
        ans += 1
    else:
        if tmp < 0 and tmp >= -k:
            ans += 1
    two *= 2
print (ans)

File: 暑期比赛专用/contest.java
import java.util.Scanner;
import java.math.BigInteger;
import java.io.BufferedInputStream;

public class Main{
	private static BigInteger two = BigInteger.valueOf(2);
	public int f(int a){
		return (a < 0 ? -a : a);
	}
	public static void main(String[] args){
		Scanner in = new Scanner(new BufferedInputStream(System.in));
		
		int n = in.nextInt();
		int k = in.nextInt();

		int val = 0;
		BigInteger sum = BigInteger.ZERO;
		BigInteger b[] = new BigInteger[n];
		b[0] = BigInteger.ONE;
		for ( int i = 0; i <= n; ++i) {
			val = in.nextInt();
			
			sum = sum.add(b[i].multiply(BigInteger.valueOf(val)));
			if(i == 0) b[i] = b[i-1].multiply(two);
		}

		int ans = 0;
		for(int i = 0; i <= n; ++i){
			if(sum.mod(b[i]).equals(BigInteger.ZERO)){
				int tmp = sum.divide(b[i]).intValue();
				if(f(tmp) <= k) ++ans;
			}
		}
		System.out.println(tmp);
	}
}